bundles/core/src/main/java/com/adobe/cq/commerce/core/components/models/retriever/AbstractProductRetriever.java

```diff
@@ -0,0 +1,216 @@
+/*******************************************************************************
+ *
+ *    Copyright 2019 Adobe. All rights reserved.
+ *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License. You may obtain a copy
+ *    of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software distributed under
+ *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ *    OF ANY KIND, either express or implied. See the License for the specific language
+ *    governing permissions and limitations under the License.
+ *
+ ******************************************************************************/
+
+package com.adobe.cq.commerce.core.components.models.retriever;
+
+import java.util.List;
+import java.util.function.Consumer;
+
+import com.adobe.cq.commerce.core.components.client.MagentoGraphqlClient;
+import com.adobe.cq.commerce.graphql.client.GraphqlResponse;
+import com.adobe.cq.commerce.magento.graphql.FilterTypeInput;
+import com.adobe.cq.commerce.magento.graphql.Operations;
+import com.adobe.cq.commerce.magento.graphql.ProductFilterInput;
+import com.adobe.cq.commerce.magento.graphql.ProductInterface;
+import com.adobe.cq.commerce.magento.graphql.ProductInterfaceQueryDefinition;
+import com.adobe.cq.commerce.magento.graphql.ProductsQueryDefinition;
+import com.adobe.cq.commerce.magento.graphql.Query;
+import com.adobe.cq.commerce.magento.graphql.QueryQuery;
+import com.adobe.cq.commerce.magento.graphql.gson.Error;
+import com.shopify.graphql.support.AbstractQuery;
+
+/**
+ * Abstract implementation of product retriever that loads product data using GraphQL.
+ */
+public abstract class AbstractProductRetriever extends AbstractRetriever {
+
+    private Consumer<AbstractQuery<?>> productQueryHook;
+    private Consumer<AbstractQuery<?>> variantQueryHook;
+    private ProductInterface product;
+    private String mediaBaseUrl;
+    private String identifier;
+
+    public AbstractProductRetriever(MagentoGraphqlClient client) {
+        super(client);
+    }
+
+    /**
+     * Returns the product.
+     *
+     * @return Product
+     */
+    public ProductInterface getProduct() {
+        if (this.product == null) {
+            populate();
+        }
+        return this.product;
+    }
+
+    /**
+     * Stores a product.
+     *
+     * @param product Product
+     */
+    protected void setProduct(ProductInterface product) {
```

Comment:
Similar comment as earlier regarding the use of `protected` fields vs. `protected` methods. It's a matter of taste so feel free to decide.